api/task_manager.cc

```diff
@@ -0,0 +1,163 @@
+                        return filter_tasks(task.second, params);
+                    });
+                    for (auto& [task_id, task] : filtered_tasks) {
+                        tm::task_stats ts;
+                        ts.task_id = task->get_status().id.to_sstring();
+                        ts.state = task->get_status().state;
+                        res.push_back(ts);
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
+        }).then([] (const std::vector<tm::task_stats>& res) {
+            return make_ready_future<json::json_return_type>(res);
+        });
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return task_manager::lookup_task_on_all_shards(ctx.tm, id).then_wrapped([&tm = ctx.tm] (auto f) {
+            if (f.failed()) {
+                auto ex = f.get_exception();
+                return make_exception_future<json::json_return_type>(std::move(ex));
+            }
+            task_manager::foreign_task_ptr task = f.get();
+            return tm.invoke_on(task.get_owner_shard(), [task = std::move(task)] (auto &tm) mutable {
+                auto state = task->get_status().state;
+                if (state == task_manager::task_state::done || state == task_manager::task_state::failed) {
+                    task->unregister();
+                }
+                return retrieve_status(std::move(task));
+            });
+        });
+    });
+
+    tm::abort_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return ctx.tm.invoke_on_all([id] (task_manager& tm) {
+            // FIXME: implement
+        }).then([]{
+             return make_ready_future<json::json_return_type>(json_void());
+        });
+    });
+
+    tm::wait_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return task_manager::lookup_task_on_all_shards(ctx.tm, id).then_wrapped([&tm = ctx.tm] (auto f) {
+            if (f.failed()) {
+                auto ex = f.get_exception();
+                return make_exception_future<json::json_return_type>(std::move(ex));
+            }
+            task_manager::foreign_task_ptr task = f.get();
+            return tm.invoke_on(task.get_owner_shard(), [t = std::move(task)] (auto& tm) mutable {
+                task_manager::task_ptr task = t.release();
+                return task->done().then_wrapped([task] (auto f) {
+                    task->unregister();
+                    if (f.failed()) {
+                        auto ex = f.get_exception();
+                        return make_exception_future<json::json_return_type>(std::move(ex));
+                    }
+                    return retrieve_status(make_foreign<task_manager::task_ptr>(task));
```

Comment:
This indicates that this needs to happen on the calling shard, not inside the invoke_on functor.